contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol

```diff
@@ -0,0 +1,105 @@
+     * @param _l2Usdc Native USDC address on L2.
+     * @param _cctpTokenMessenger TokenMessenger contract to bridge via CCTP.
+     * @param _customGasTokenFunder Contract that funds the custom gas token.
+     * @param _l3MaxSubmissionCost Amount of gas token allocated to pay for the base submission fee. The base
+     * submission fee is a parameter unique to Arbitrum retryable transactions. This value is hardcoded
+     * and used for all messages sent by this adapter.
+     */
+    constructor(
+        ArbitrumCustomGasTokenInbox _l2ArbitrumInbox,
+        ArbitrumL1ERC20GatewayLike _l2ERC20GatewayRouter,
+        address _l3RefundL3Address,
+        IERC20 _l2Usdc,
+        ICCTPTokenMessenger _cctpTokenMessenger,
+        uint32 _circleDomainId,
+        FunderInterface _customGasTokenFunder,
+        uint256 _l3MaxSubmissionCost,
+        uint256 _l3GasPrice,
+        address _l3SpokePool,
+        address _crossDomainAdmin
+    )
+        Arbitrum_CustomGasToken_AdapterBase(
+            _l2ArbitrumInbox,
+            _l2ERC20GatewayRouter,
+            _l3RefundL3Address,
+            _l2Usdc,
+            _cctpTokenMessenger,
+            _circleDomainId,
+            _customGasTokenFunder,
+            _l3MaxSubmissionCost,
+            _l3GasPrice
+        )
+        ForwarderBase()
+    {}
+
+    /**
+     * @notice Bridge tokens to Arbitrum-like L3.
+     * @notice This contract must hold at least getL2CallValue() amount of ETH or custom gas token
+     * to send a message via the Inbox successfully, or the message will get stuck.
+     * @param l2Token L2 token to send.
+     * @param amount Amount of L2 tokens to deposit and L3 tokens to receive.
+     */
+    function relayTokens(
+        address l2Token,
+        address,
+        uint256 amount
+    ) external payable override {
+        _relayTokens(l2Token, address(0), amount, l3SpokePool);
+        emit TokensForwarded(l2Token, amount);
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Arbitrum-like L3.
+     * @notice This contract must hold at least getL2CallValue() amount of the custom gas token
+     * to send a message via the Inbox successfully, or the message will get stuck.
+     * @param target Contract on Arbitrum-like L3 that will receive message.
+     * @param message Data to send to target.
+     */
+    function _relayL3Message(address target, bytes memory message) internal override {
+        _relayMessage(target, message);
+        emit MessageForwarded(target, message);
+    }
+
+    function _requireAdminSender() internal virtual override onlyFromCrossDomainAdmin {}
+
+    function _applyL1ToL2Alias(address l1Address) internal pure returns (address l2Address) {
+        // Allows overflows as explained above.
+        unchecked {
+            l2Address = address(uint160(l1Address) + uint160(0x1111000000000000000000000000000000001111));
+        }
+    }
```

Comment:
We have this in the SpokePool as well. Should we factor it out to an "Arbitrum Utils" library for inlining at build time?

Reply:
Same comment as paul. If we add it as an internal library function

Reply:
Refactored: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b